boutdata/collect.py

```diff
@@ -307,392 +236,776 @@ def getDataFile(i):
         if yguards == "include_upper" and f["jyseps2_1"] != f["jyseps1_2"]:
             # Simulation has a second (upper) target, with a second set of y-boundary
             # points
-            ny = ny + 2*myg
-            ny_inner = f["ny_inner"]
-            yproc_upper_target = ny_inner // mysub - 1
-            if f["ny_inner"] % mysub != 0:
-                raise ValueError("Trying to keep upper boundary cells but "
-                                 "mysub={} does not divide ny_inner={}"
-                                 .format(mysub, ny_inner))
-        else:
-            yproc_upper_target = None
+            ny = ny + 2 * myg
     else:
-        ny = mysub * nype
+        ny = f["ny"]
+    nz = f["MZ"]
+    t_array = f.read("t_array")
+    if t_array is None:
+        nt = 1
+        t_array = np.zeros(1)
+    else:
+        try:
+            nt = len(t_array)
+        except TypeError:
+            # t_array is not an array here, which probably means it was a
+            # one-element array and has been read as a scalar.
+            nt = 1
 
     xind = _convert_to_nice_slice(xind, nx, "xind")
     yind = _convert_to_nice_slice(yind, ny, "yind")
     zind = _convert_to_nice_slice(zind, nz, "zind")
     tind = _convert_to_nice_slice(tind, nt, "tind")
 
-    xsize = xind.stop - xind.start
-    ysize = yind.stop - yind.start
-    zsize = int(np.ceil(float(zind.stop - zind.start)/zind.step))
-    tsize = int(np.ceil(float(tind.stop - tind.start)/tind.step))
+    if not xguards:
+        xind = slice(xind.start + mxg, xind.stop + mxg, xind.step)
+    if not yguards:
+        yind = slice(yind.start + myg, yind.stop + myg, yind.step)
+
+    if dimensions == ():
+        ranges = []
+    elif dimensions == ("t",):
+        ranges = [tind]
+    elif dimensions == ("x", "y"):
+        # Field2D
+        ranges = [xind, yind]
+    elif dimensions == ("x", "z"):
+        # FieldPerp
+        ranges = [xind, zind]
+    elif dimensions == ("t", "x", "y"):
+        # evolving Field2D
+        ranges = [tind, xind, yind]
+    elif dimensions == ("t", "x", "z"):
+        # evolving FieldPerp
+        ranges = [tind, xind, zind]
+    elif dimensions == ("x", "y", "z"):
+        # Field3D
+        ranges = [xind, yind, zind]
+    elif dimensions == ("t", "x", "y", "z"):
+        # evolving Field3D
+        ranges = [tind, xind, yind, zind]
+    else:
+        # Not a Field, so do not support slicing.
+        # For example, may be a string.
+        ranges = None
+
+    data = f.read(varname, ranges)
+    var_attributes = f.attributes(varname)
+    return BoutArray(data, attributes=var_attributes)
+
+
+def _collect_from_one_proc(
+    i,
+    datafile,
+    varname,
+    *,
+    result,
+    is_fieldperp,
+    grid_info,
+    dimensions,
+    tind,
+    xind,
+    yind,
+    zind,
+    xguards,
+    yguards,
+    info,
+    parallel_read=False,
+):
+    """Read part of a variable from one processor
+
+    For use in _collect_parallel()
```

Comment:
Maybe mention that the result is not returned?
I managed to confuse myself when reading the code as I expected the data to be returned as a parameter (rather than modify a mutable input parameter)